:seedling: Enable exhaustive linter

erjavaskivuori opened the pull request:
<!-- please add a icon to the title of this PR and delete this line and similar ones -->
<!-- the icon will be either ⚠️ (:warning:, major or breaking changes), ✨ (:sparkles:, feature additions), 🐛 (:bug:, patch and bugfixes), 📖 (:book:, documentation or proposals), or 🌱 (:seedling:, minor or other) -->

**What this PR does / why we need it**: Enable exhaustive linter to check exhaustiveness of switch statements of enum-like constants. Those having only one case are changed into an if-statement. 

**Which issue(s) this PR fixes**:
Fixes #2355

tuminoid commented:
Testing copilot review here, let's see what it hallucinates :)

dtantsur commented:
I think the AI is correct about the issue.

Also, I'm not entirely fond of what this lint is making is do.. I think in many places the switches were somewhat more readable. Also, depending how good the Go's optimizer is, the new code is significantly less performant.

erjavaskivuori commented:
Yes, I need to take closer a look, it seems that I have missed something there actually.

I was debating what approach to take here. The linter wants that for example every possible state is covered in switch statements. In most if not all  there were +10 cases missing and just listing those for no op also feels a bit funny to me. With the configuration I added, `default:` can be used instead of actually listing everything. I used that then in some cases but in some I changed the switch to if-statement if there were like only 3 out of 20 states covered initially. But as you mentioned, there is a possibility that it affects the performance. 

Do you have an opinion should I rather keep them as they were and use `default:` (or list the missing cases) to satisfy the linter? @dtantsur

dtantsur commented:
I'm fine with using an `if` statement in simplest cases, but when we feel the need to use a slice with expected values (3+ possibilities?), I think it's better to add a `default:` branch with nothing (or a comment why nothing is happening).

erjavaskivuori commented:
Okay good, I will do the changes accordingly. I was intrigued to find something similar to `if x in list` as in Python and hence the slices but maybe there the default with no op can be better then :smile:

tuminoid commented on apis/metal3.io/v1alpha1/baremetalhost_types.go:
```diff
@@ -1115,6 +1115,7 @@ func (host *BareMetalHost) OperationMetricForState(operation ProvisioningState)
 		metric = &history.Provision
 	case StateDeprovisioning:
 		metric = &history.Deprovision
+	default:
```
In the terms of the exhaustive linter, Copilot has a point, but OTOH we're also enabling the config to accept empty defaults for exhaustive, so :shrug:  At least it labeled it as nitpick, so not too controvesial comment I guess.

tuminoid commented on pkg/provisioner/ironic/ironic.go:
```diff
@@ -338,31 +339,24 @@ func (p *ironicProvisioner) configureImages(data provisioner.ManagementAccessDat
+	if data.State == metal3api.StateDeprovisioning && data.AutomatedCleaningMode == metal3api.CleaningModeDisabled {
+		// No need for pre-provisioning image if cleaning disabled
+		return result, nil
+	}
```
> (Also, what's wrong with a switch statement with only 1 case + empty default? The if statement is both less readable IMHO and harder to modify.)

It is just what exhaustive linter wants, unless you have the "accept default too" option turned on. It is supposed to prevent unintentional falling back to default.

dtantsur commented:
The goal is to be explicit in your intentions and avoid omitting values accidentally. We could actually list all values instead of `default:` to make sure whenever we expand the list of states we also update these switch blocks. I wonder if it's worth the effort here. Using `default:` might be good enough.

Using a slice, however, is hiding the issue instead of fixing it. You have the same non-exhaustive list, just the linter now does not understand it. The more I think about it, the more I'm against using slices instead of switches.

tuminoid reviewed (commented):
/lgtm
